### ISSUE TITLE
system_rng silently depends on cryptoapi_rng on Win10

### ISSUE THREAD

**Issue by AsderYork** [randombit/botan#1078]
If configured with "configure.py --cc=msvc --minimized-build --enable-modules=rng,system_rng,sha2_32,bcrypt,auto_rng --amalgamation", nmake fails with errors on std::min/max. After adding #define NOMINMAX at the top of amalgamation header, build still fails on linker stage with LNK2019 on functions from Wincrypt.h. After changing configuration to "configure.py --cc=msvc --minimized-build --enable-modules=rng,system_rng,sha2_32,bcrypt,auto_rng,cryptoapi_rng --amalgamation" build have succeed.

**Comment by noloader** [comment]
Related, see [PR 740, define NOMINMAX 1 before every include of windows.h](https://github.com/randombit/botan/pull/740#issuecomment-285329517).

**Comment by webmaster128** [comment]
Is there a good reason why we have

```
#include <windows.h>
#define NOMINMAX 1
#include <wincrypt.h>
```
instead of 

```
#define NOMINMAX 1
#include <windows.h>
#include <wincrypt.h>
```

in src/lib/rng/system_rng/system_rng.cpp? This looks like a mistake to me in #740. cc @BMBurstein

------------------------

@AsderYork Could you try adding

```
<libs>
windows -> crypt32.lib
mingw -> crypt32
</libs>
```

or 

```
<libs>
windows -> advapi32.lib
mingw -> advapi32
</libs>
```

to src/lib/rng/system_rng/info.txt? Because of

> Linking CryptoAPI functions requires the availability of Crypt32.lib. Programs using some of the functions may also require Advapi32.lib. When using the Visual C++ compiler, these files should be listed among the object/library modules.

https://msdn.microsoft.com/en-us/library/windows/desktop/aa387673(v=vs.85).aspx

In src/lib/entropy/cryptoapi_rng/info.txt, we link advapi32.lib, which is why adding cryptoapi_rng to the configuration helped in your case.

**Comment by BMBurstein** [comment]
> Is there a good reason why we have
> 
> #include <windows.h>
> #define NOMINMAX 1
> #include <wincrypt.h>
> instead of
> 
> #define NOMINMAX 1
> #include <windows.h>
> #include <wincrypt.h>
> in src/lib/rng/system_rng/system_rng.cpp? This looks like a mistake to me in #740. cc @BMBurstein

Sorry, definitely my mistake

**Comment by webmaster128** [comment]
@BMBurstein, could you open a PR patching the NOMINMAX issue?

**Comment by randombit** [comment]
Thanks for the report. This should now be fixed in master (and in upcoming 2.3.0 release), as `cryptoapi_rng` module has been removed and `system_rng` handles everything for accessing `CryptGenRandom` and company.
